### ISSUE TITLE
Silly question: I'm probably doing this wrong right? :-)

### ISSUE THREAD

**Issue by jliebrand** [mapbox/tippecanoe#493]
Hey, we use tippecanoe for a lot of the data that we put together and it's been great. So when I recently wanted to add a visualisation to our product, I didn't think twice and just built out the same approach I've done before, but I now realise I might be an idiot for doing this. So advice would be great.

Basically, I'm trying to build a gradient layer in mapbox that represents temperatures across the globe. Either using the new heatmap layer, or (for native where that is not yet supported?), I simply draw a circle layer, with the radius of the circle big enough for them to slightly overlap, and then set blur to 1. This produces a nice gradient effect.

However, for the circles radius to work at various zoom levels, I need different granular data at those levels. At the moment, I'm producing a grid as follows:

Zoom 0-2:  a grid of points every 5degrees
Zoom 2-8: a grid of points every 0.25 degrees
Zoom 9: a grid of points every 0.025 degrees

Clearly the dataset at zoom 9 is huge for the entire globe (11GB GeoJSON; 103 million feature points). But once converted to an mbtileset it compresses down to 1.6GB which is more manageable.

To go from my GeoJSON to mbtileset I'm using tippecanoe with -z9 -Z9 -B9.  In other words, for zoom level 9 do not lose any points. So all in all this "works"... This takes HOURS to run!

And then it occurred to me that since I'm not doing *any* point simplification at all, should I really be using tippecanoe? Is there a more obvious way to go from geoJSON to mbtileset for this case?

In short, I'm I being a pillock here? :-)

Is there a better way to convert the data; or even better still, is there a better way to create a gradient layer for which I do not need as many data points (the 0.025 grid is simply a bi-linear interpolation of the 0.25 grid after all).

**Comment by jliebrand** [comment]
To be even more precise, I run tippecanoe three times for zoom 0-2, 3-8 and zoom 9. I then subsequently run tile-join.

At the moment, it's *nearly* done with tile-join and it's been going for 18 hours :-)

**Comment by ericfischer** [comment]
You're right, gridded data is not something that Tippecanoe handles very well. [Tile-count](https://github.com/mapbox/tile-count) solves one specific case of the problem, but probably not your use case. I'll try to think of a good way to make these sorts of tiles.

**Comment by ericfischer** [comment]
After giving this some more attention, I think Tippecanoe actually works reasonably well for this at z9, although I wouldn't want to push it too many zoom levels deeper.

I don't have your temperature data, so instead I tried [making a worldwide gridded elevation tileset](https://github.com/ericfischer/srtm-gridded-vector) at the same 0.025 degree resolution at z9. I'm not sure why it is running so slowly for you, because it only took a little over an hour on my (4-CPU, MacBook Air) laptop.

The one thing I am definitely doing differently is that I am using the `"tippecanoe": { "minzoom": …, "maxzoom": … }` extension to specify each feature's zoom level rather than doing multiple runs and merging them together. But I wouldn't think that would make much difference in the overall running time.

If your data is already structured into web map tiles, you might be able to do better with some other tool that can take advantage of that, but I can't think of anything immediately that would do especially better with a degree-oriented data grid.

**Comment by jliebrand** [comment]
And not surprisingly it turns out I have indeed been a pillock :-)

I was sure I was ssh'd in to a decent machine when I ran this, but a little copy'n'paste error on the ip address meant that I was actually ssh'd in to a t2.micro machine.... no wonder it took that long!

Running this on a decent machine worked much better. Which reminds me: for a large dataset like this, what is actually the restricting factor? CPU (aka number of cores), memory or IO ?

I'll run my dataset on a handful of different EC2 images so I can compare numbers, but thought I'd ask.

ps. good point about using the extension: I actually think that *can* save quite a bit, because the join takes quite a long time as well. I'll run some comparison numbers on that also...

**Comment by ericfischer** [comment]
Glad to hear there was a reasonable explanation!

If there are a lot of unique feature attributes, memory can sometimes be a bottleneck because it builds a tree of attribute values for deduplication. Beyond that, you need to have enough memory (or at least swap space) to keep the uncompressed data for as many tiles as you have CPUs in memory at once during tiling.

More commonly, CPU is the bottleneck, because there is a lot of work to be done for each tile. Other than at the lowest zoom levels many tiling threads can run at the same time, so the more CPUs you have, the better.

I/O will mostly be a bottleneck during the "reordering geometry" phase between input and tiling, where it reads through the entire geometry and metadata structure that it has parsed and writes it out again in a different order. It also has to read and rewrite all that data for each zoom level during tiling, so it might have to do physical I/O instead of working entirely within the buffer cache, but there is enough CPU work going on with the tiling that it's usually not waiting for the disk.
